File: Codes/003_lengthOfLongestSubstring/3.无重复字符的最长子串.js
/*
 * @Author: Theo_hui
 * @Date: 2021-08-17 10:59:52
 * @Descripttion: 
 */
/*
 * @lc app=leetcode.cn id=3 lang=javascript
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let res=0, left=-1
    let c_maps = new Array()

    for(var right=0;right<s.length;right++){
        c_i=c_maps[s[right]]

        if(typeof(c_i) != "undefined" && c_i>left){
            left = c_i
        }else{
            let new_res = right-left
            if(new_res>res){
                res=new_res
            }
        }
        c_maps[s[right]]=right
    }

    return res
};
// @lc code=end

File: Codes/003_lengthOfLongestSubstring/3.无重复字符的最长子串.go
/*
 * @Author: Theo_hui
 * @Date: 2021-08-17 10:45:55
 * @Descripttion:
 */
/*
 * @lc app=leetcode.cn id=3 lang=golang
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	res, left := 0, -1           //结果，左下标
	c_maps := make(map[rune]int) // 字符最右下标字典

	for right, c := range s {
		i, ok := c_maps[c] //该字符的最右下标

		if ok && i > left {
			//字符重复了
			left = i
		} else {
			//没有重复，更新结果
			new_res := right - left
			if new_res > res {
				res = new_res
			}
		}
		// 更新字符最右位置
		c_maps[c] = right
	}
	return res
}

// @lc code=end

